src/IO/AzureBlobStorage/copyAzureBlobStorageFile.cpp

```diff
@@ -0,0 +1,407 @@
+    extern const Event DiskAzureUploadPart;
+}
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int INVALID_CONFIG_PARAMETER;
+    extern const int AZURE_BLOB_STORAGE_ERROR;
+}
+
+
+size_t max_single_operation_copy_size = 256 * 1024 * 1024;
+
+
+namespace
+{
+    class UploadHelper
+    {
+    public:
+        UploadHelper(
+            const CreateReadBuffer & create_read_buffer_,
+            MultiVersion<Azure::Storage::Blobs::BlobContainerClient> & client_,
+            size_t offset_,
+            size_t total_size_,
+            const String & dest_container_,
+            const String & dest_blob_,
+            std::shared_ptr<AzureObjectStorageSettings> settings_,
+            const std::optional<std::map<String, String>> & object_metadata_,
+            ThreadPoolCallbackRunner<void> schedule_,
+            bool for_disk_azure_blob_storage_,
+            const Poco::Logger * log_)
+            : create_read_buffer(create_read_buffer_)
+            , client(client_)
+            , offset (offset_)
+            , total_size (total_size_)
+            , dest_container(dest_container_)
+            , dest_blob(dest_blob_)
+            , settings(settings_)
+            , object_metadata(object_metadata_)
+            , schedule(schedule_)
+            , for_disk_azure_blob_storage(for_disk_azure_blob_storage_)
+            , log(log_)
+            , max_single_part_upload_size(settings_->max_single_part_upload_size)
+        {
+        }
+
+        virtual ~UploadHelper() = default;
+
+    protected:
+        std::function<std::unique_ptr<SeekableReadBuffer>()> create_read_buffer;
+        MultiVersion<Azure::Storage::Blobs::BlobContainerClient> & client;
+        size_t offset;
+        size_t total_size;
+        const String & dest_container;
+        const String & dest_blob;
+        std::shared_ptr<AzureObjectStorageSettings> settings;
+        const std::optional<std::map<String, String>> & object_metadata;
+        ThreadPoolCallbackRunner<void> schedule;
+        bool for_disk_azure_blob_storage;
+        const Poco::Logger * log;
+        size_t max_single_part_upload_size;
+
+        struct UploadPartTask
+        {
+            char *data = nullptr;
+            size_t size = 0;
+            std::string block_id;
+            bool is_finished = false;
+            std::exception_ptr exception;
+
+            ~UploadPartTask()
+            {
+                if (data != nullptr)
+                    free(data);
+            }
+        };
+
+        size_t normal_part_size;
+        std::vector<std::string> block_ids;
+
+        std::list<UploadPartTask> TSA_GUARDED_BY(bg_tasks_mutex) bg_tasks;
+        int num_added_bg_tasks TSA_GUARDED_BY(bg_tasks_mutex) = 0;
+        int num_finished_bg_tasks TSA_GUARDED_BY(bg_tasks_mutex) = 0;
+        std::mutex bg_tasks_mutex;
+        std::condition_variable bg_tasks_condvar;
+
+        void calculatePartSize()
+        {
```

Comment:
This function is too complicated for Azure. The following should be enough:

```
void calculatePartSize()
{
    auto max_upload_part_size = settings->max_upload_part_size;
    if (!max_upload_part_size)
        throw Exception(ErrorCodes::INVALID_CONFIG_PARAMETER, "max_upload_part_size must not be 0");
    /// We've calculated the size of a normal part (the final part can be smaller).
    normal_part_size = max_upload_part_size;
}
```

Thus instead of `max_part_number`, `min_upload_part_size`, `max_upload_part_size` we only need `max_upload_part_size`.

Reply:
addressed it